### PR TITLE
WordAds: Swap ad calls to dynamic

### PR DESCRIPTION
**Pull request by dbspringer**
Our partners at IPW have developed a more dynamic ad loader that allows us to more generally describe the space the ad is in and find the best unit available. This code replaces the standard units (header/belowpost/sidebar) with the dynamic ads. Units added through AMP or Gutenberg will retain their original loaders as there's more work required to make them compliant.

#### Changes proposed in this Pull Request:

* Updates __ATA ad inits to utilize the dynamic version.
* Units added through Gutenberg or AMP will retain their original calls.

#### Does this pull request change what data or activity we track or use?
* No

#### Testing instructions:

* Notify @dbspringer of the site you plan on testing so he can make sure the dynamic ads are enabled for your site in the backend.
* Enable the Ads module in Jetpack Settings -> Traffic on a Jetpack test site with Premium+ plan.
* Enable all the ad units, some widgets, and create a post with a few Gutenberg ad blocks.
* Ads should load as you'd expect, the inserted ad snippets should look something like:
```
<script>
	__ATA.cmd.push(function() {
		__ATA.initDynamicSlot({
			id: 'atatags-1576465421-5efd1825ef585',
			location: 120,
			formFactor: '001',
			label: {
				text: 'Advertisements',
			},
			creative: {
				reportAd: {
					text: 'Report this ad',
				},
				privacySettings: {
					text: 'Privacy settings',
				}
			}
		});
	});
</script>
```

#### Proposed changelog entry for your changes:

* Updating ad units to be more flexible and dynamic

### CODE REVIEW

File: modules/wordads/class-wordads.php [hunk 1]
```diff
@@ -625,51 +635,120 @@ public function get_ad( $spot, $type = 'iponweb' ) {
...
+
+		return $this->get_dynamic_ad_snippet( $section_id, $form_factor, $location );
+	}
+
+	/**
+	 * Returns the dynamic snippet to be inserted into the ad unit
+	 *
+	 * @param  int    $section_id  section_id.
+	 * @param  string $form_factor form_factor.
+	 * @param  int    $location    location.
```
Comment (cphilleo) [inline modules/wordads/class-wordads.php:692]: `$location` param is a `string`.

Comment (dbspringer) [reply modules/wordads/class-wordads.php:692]: Derp, good catch.

**Comment by dbspringer** [comment]
@kraftbj @jeherve since the two of you are generally the ones that help out with our PR reviews can you let me know what your testing sites are and I'll make sure the dynamic ads get turned on your your sites.

**Comment by jeherve** [comment]
@dbspringer Thank you. Mine can be `176738923`.

**Comment by kraftbj** [comment]
Thanks! Let's use `23454191` for me.

File: modules/wordads/class-wordads.php [hunk 2]
```diff
@@ -625,51 +635,120 @@ public function get_ad( $spot, $type = 'iponweb' ) {
 	 * @since 5.7
 	 */
 	public function get_ad_snippet( $section_id, $height, $width, $location = '', $css = '' ) {
+		if ( function_exists( 'is_amp_endpoint' ) && is_amp_endpoint() ) {
```
Comment (jeherve) [inline modules/wordads/class-wordads.php:638]: Let's use `class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()` instead, to be consistent with other places in Jetpack.

File: modules/wordads/class-wordads.php [hunk 3]
```diff
@@ -625,51 +635,120 @@ public function get_ad( $spot, $type = 'iponweb' ) {
...
-			$amp_section_id = esc_attr( self::get_amp_section_id() );
-			$site_id        = esc_attr( $this->params->blog_id );
+		if ( 'gutenberg' === $location ) {
+			$ad_number = count( $this->ads ) . '-' . uniqid();
+			$data_tags = $this->params->cloudflare ? ' data-cfasync="false"' : '';
+			$css       = esc_attr( $css );
+
+			$loc_id = 100;
+			if ( ! empty( self::$ad_location_ids[ $location ] ) ) {
+				$loc_id = self::$ad_location_ids[ $location ];
```
Comment (jeherve) [inline modules/wordads/class-wordads.php:655]: Should we escape this?

File: modules/wordads/class-wordads.php [hunk 4]
```diff
@@ -625,51 +635,120 @@ public function get_ad( $spot, $type = 'iponweb' ) {
...
+	 * @since 8.7
+	 */
+	public function get_dynamic_ad_snippet( $section_id, $form_factor = 'square', $location = '', $relocate = '' ) {
+		$div_id = 'atatags-' . $section_id . '-' . uniqid();
+		$div_id = esc_js( $div_id );
+
+		// Default form factor.
+		$form_factor_id = self::$form_factor_ids['square'];
+		if ( isset( self::$form_factor_ids[ $form_factor ] ) ) {
+			$form_factor_id = self::$form_factor_ids[ $form_factor ];
```
Comment (jeherve) [inline modules/wordads/class-wordads.php:705]: Does this need to be escaped?

**Pull request merged**